File: src/lang/parse/literals.zig
const std = @import("std");
const compstate = @import("../compiler_state.zig");
const token = @import("token.zig");
const util = @import("../../util.zig");
const keywords = @import("keywords.zig");

pub fn read_identifier(c: *compstate.CompilerState) void {
    var id = "";
    var nextc = c.file[c.lex_i + 1];

    while (std.ascii.isAlphanumeric(nextc) || nextc == '_') {
        id = util.strconcat(id, c.file[c.lex_i]);
        c.lex_i += 1;
        nextc = c.file[c.lex_i + 1];
    }

    const nowthatlookslikeaj = keywords.read_keyword(id);
    switch (nowthatlookslikeaj) {
        .identifier => c.tokens.append(token.Token { .identifier = id }),
        else => c.tokens.append(nowthatlookslikeaj),
    }
}

pub fn read_string(c: *compstate.CompilerState) void {
    var str = "";
    var nextc = c.file[c.lex_i + 1];

    // TODO: add escapes like \n \" \u69420
    while (nextc != '"') {
        str = util.strconcat(str, c.file[c.lex_i]);
        c.lex_i += 1;
        nextc = c.file[c.lex_i + 1];
    }
}

pub fn read_number(c: *compstate.CompilerState) void {
    var is_float = false;
    var numstr = "";
    var nextc = c.file[c.lex_i + 1];

    while (std.ascii.isDigit(nextc)) {
        numstr = util.strconcat(numstr, c.file[c.lex_i]);
        if (nextc == '.') {
            c.lex_i += 1;
            is_float = true;
        }
        c.lex_i += 1;
        nextc = c.file[c.lex_i + 1];
    }

    // we just parsed a string, strings are in fact not numbers
    // conveniently parseInt/parseFloat supports _s like 1_000_000
    if (is_float) {
        return try std.fmt.parseFloat(f64, numstr);
    }
    else {
        return try std.fmt.parseInt(i64, numstr, 10);
    }
}

File: src/lang/lexer/Literals.cs
using System;
using System.Globalization;
namespace kettlevm;

public static partial class Lexer {
    static CompilerState read_identifier(ref CompilerState c)
    {
        string id = "";

        while (is_alphabetic(c.nextc()) || is_digit(c.nextc()) || c.nextc() == '_') {
            id += c.thisc();
            c.lex_i++;
        }

        // read_keyword returns an id token if it's not a keyword
        Token nowthatlookslikeaj = read_keyword(id);
        c.tokens.Enqueue(nowthatlookslikeaj.type switch {
            TokenType.identifier => new(TokenType.identifier) { strval = id },
            _ => nowthatlookslikeaj
        });
        return c;
    }

    static CompilerState read_string(ref CompilerState c)
    {
        string str = "";

        while (c.nextc() != '"') {
            // escape stuff
            if (c.nextc() == '\\') {
                c.lex_i++;
                str += c.thisc();
            }
            else {
                str += c.thisc();
            }
            c.lex_i++;
        }

        c.tokens.Enqueue(new(TokenType.strlit) { strval = str });
        return c;
    }

    static CompilerState read_char(ref CompilerState c)
    {
        string str = "";

        while (c.nextc() != '\'') {
            // escape stuff
            if (c.nextc() == '\\') {
                c.lex_i++;
                str += c.thisc();
            }
            else {
                str += c.thisc();
            }
            c.lex_i++;
        }

        if (str.Length == 1) c.tokens.Enqueue(new(TokenType.charlit) { strval = str });
        else c.complain("Characters must be 1 character");
        return c;
    }

    static CompilerState read_number(ref CompilerState c)
    {
        bool isfloat = false;
        string numstr = "";
        
        // _ is supported so you can write really big numbers like 1_000_000
        while (is_digit(c.nextc()) || c.nextc() == '_') {
            numstr += c.thisc();
            if (c.nextc() == '.') {
                c.lex_i++;
                isfloat = true;
            }
            c.lex_i++;
        }

        // we just parsed a string, strings are in fact not numbers
        numstr = numstr.Replace("_", "");
        if (isfloat) {
            c.tokens.Enqueue(new(TokenType.floatlit) { floatval = double.Parse(numstr, CultureInfo.InvariantCulture) });
        }
        else {
            c.tokens.Enqueue(new(TokenType.intlit) { intval = ulong.Parse(numstr, CultureInfo.InvariantCulture) });
        }
        return c;
    }
}